ndarray/src/main/java/org/tensorflow/ndarray/impl/sparse/AbstractSparseNdArray.java

```diff
@@ -0,0 +1,481 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.ndarray.impl.sparse;
+
+import org.tensorflow.ndarray.IllegalRankException;
+import org.tensorflow.ndarray.LongNdArray;
+import org.tensorflow.ndarray.NdArray;
+import org.tensorflow.ndarray.NdArraySequence;
+import org.tensorflow.ndarray.NdArrays;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.ndarray.SparseNdArray;
+import org.tensorflow.ndarray.StdArrays;
+import org.tensorflow.ndarray.impl.AbstractNdArray;
+import org.tensorflow.ndarray.impl.dense.AbstractDenseNdArray;
+import org.tensorflow.ndarray.impl.dimension.Dimension;
+import org.tensorflow.ndarray.impl.dimension.DimensionalSpace;
+import org.tensorflow.ndarray.impl.dimension.RelativeDimensionalSpace;
+import org.tensorflow.ndarray.impl.sequence.SingleElementSequence;
+import org.tensorflow.ndarray.impl.sequence.SlicingElementSequence;
+import org.tensorflow.ndarray.index.Index;
+
+import java.nio.ReadOnlyBufferException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.LongStream;
+
+/**
+ * Abstract base class for sparse array.
+ *
+ * <p>A sparse array as two separate dense arrays: indices, values, and a shape that represents the
+ * dense shape.
+ *
+ * <p><em>NOTE:</em> all Sparse Arrays are readonly for the {@link #set(NdArray, long...)} and
+ * {@link #setObject(Object, long...)} methods
+ *
+ * <pre>{@code
+ * FloatSparseNdArray st = new FloatSparseNdArray(
+ *      StdArrays.of(new long[][] {{0, 0}, {1, 2}},
+ *      StdArrays.of(new float[] {1f, 2f},
+ *      Shape.of(3, 4));
+ *
+ * }</pre>
+ *
+ * <p>represents the dense array:
+ *
+ * <pre>{@code
+ * [[1, 0, 0, 0]
+ *  [0, 0, 2, 0]
+ *  [0, 0, 0, 0]]
+ *
+ * }</pre>
+ *
+ * @param <T> the type that the array contains
+ * @param <U> the type of dense NdArray
+ */
+public abstract class AbstractSparseNdArray<T, U extends NdArray<T>> extends AbstractNdArray<T, U>
+    implements SparseNdArray<T, U> {
+  /**
+   * A 2-D long array of shape {@code [N, ndims]}, that specifies the indices of the elements in the
+   * sparse array that contain nonzero values (elements are zero-indexed).
+   *
+   * <p>For example, {@code indices=[[1,3], [2,4]]} specifies that the elements with indexes of
+   * coordinates {@code [1,3]} and {@code [2,4]} have nonzero values.
+   */
+  private LongNdArray indices;
+  /**
+   * A 1-D array of any type and shape {@code [N]}, that supplies the values for each element in
+   * indices.
+   *
+   * <p>For example, given {@code indices=[[1,3], [2,4]]}, and {@code values=[18, 3.6]} specifies
+   * that element {@code [1,3]} of the sparse array has a value of {@code 18}, and element {@code
+   * [2,4]} of the sparse array has a value of {@code 3.6}.
+   */
+  private U values;
```

Comment:
Nit: maybe add some spacing between the javadoc and the previous field? Same comment applies overall.

Reply:
OK